File: tests/oracle/decompress_1204.cc
#include "compression.hh"

#include <cstring>
#include <iomanip>
#include <iostream>
#include <string>

namespace {

using namespace ghidra;

const uint1 HELLO_ZLIB[] = {0x78, 0x9c, 0xcb, 0x48, 0xcd, 0xc9, 0xc9,
                            0x07, 0x00, 0x06, 0x2c, 0x02, 0x15};
const uint1 WORLD_ZLIB[] = {0x78, 0x9c, 0x2b, 0xcf, 0x2f, 0xca, 0x49,
                            0x01, 0x00, 0x06, 0xa6, 0x02, 0x29};

const char *truth(bool value) { return value ? "true" : "false"; }

void print_hex(const uint1 *buffer, int4 size) {
  std::ios_base::fmtflags flags = std::cout.flags();
  for (int4 i = 0; i < size; ++i)
    std::cout << std::hex << std::setfill('0') << std::setw(2)
              << static_cast<unsigned int>(buffer[i]);
  std::cout.flags(flags);
}

} // namespace

int main() {
  std::cout << "zlib.version=" << zlibVersion() << '\n';

  {
    Decompress stream;
    uint1 output[8] = {};
    int4 remaining = stream.inflate(output, sizeof(output));
    std::cout << "no_input.remaining=" << remaining
              << ":finished=" << truth(stream.isFinished()) << '\n';
  }

  {
    uint1 hello[sizeof(HELLO_ZLIB)];
    uint1 world[sizeof(WORLD_ZLIB)];
    std::memcpy(hello, HELLO_ZLIB, sizeof(hello));
    std::memcpy(world, WORLD_ZLIB, sizeof(world));
    Decompress stream;
    stream.input(hello, sizeof(hello));
    uint1 first[1] = {};
    int4 first_remaining = stream.inflate(first, sizeof(first));
    stream.input(world, sizeof(world));
    uint1 second[64] = {};
    int4 second_remaining = stream.inflate(second, sizeof(second));
    std::cout << "midstream_replace.first_remaining=" << first_remaining
              << ":second_remaining=" << second_remaining
              << ":finished=" << truth(stream.isFinished()) << ":hex=";
    print_hex(second, sizeof(second) - second_remaining);
    std::cout << '\n';
  }

  {
    uint1 hello[sizeof(HELLO_ZLIB)];
    std::memcpy(hello, HELLO_ZLIB, sizeof(hello));
    Decompress stream;
    stream.input(hello, sizeof(hello));
    hello[0] = 0;
    uint1 output[8] = {};
    try {
      stream.inflate(output, sizeof(output));
      std::cout << "alias_mutation.error=none\n";
    } catch (const LowlevelError &error) {
      std::cout << "alias_mutation.error=" << error.explain << '\n';
    }
  }

  {
    uint1 buffer[64] = {};
    std::memcpy(buffer, HELLO_ZLIB, sizeof(HELLO_ZLIB));
    Decompress stream;
    stream.input(buffer, sizeof(HELLO_ZLIB));
    int4 remaining = stream.inflate(buffer, sizeof(HELLO_ZLIB));
    std::cout << "same_address.remaining=" << remaining
              << ":finished=" << truth(stream.isFinished()) << ":hex=";
    print_hex(buffer, sizeof(HELLO_ZLIB) - remaining);
    std::cout << '\n';
  }

  {
    Decompress stream;
    stream.input(const_cast<uint1 *>(HELLO_ZLIB), sizeof(HELLO_ZLIB));
    uint1 first[1] = {};
    int4 first_remaining = stream.inflate(first, sizeof(first));
    std::cout << "stream.first=" << std::string((char *)first, 1)
              << ":remaining=" << first_remaining
              << ":finished=" << truth(stream.isFinished()) << '\n';
    uint1 second[64] = {};
    int4 second_remaining = stream.inflate(second, sizeof(second));
    std::cout << "stream.second="
              << std::string((char *)second, sizeof(second) - second_remaining)
              << ":remaining=" << second_remaining
              << ":finished=" << truth(stream.isFinished()) << '\n';
  }

  {
    uint1 invalid[] = {0xde, 0xad, 0xbe, 0xef};
    Decompress stream;
    stream.input(invalid, sizeof(invalid));
    uint1 output[8] = {};
    try {
      stream.inflate(output, sizeof(output));
      std::cout << "invalid.error=none\n";
    } catch (const LowlevelError &error) {
      std::cout << "invalid.error=" << error.explain
                << ":finished=" << truth(stream.isFinished()) << '\n';
    }
  }

  {
    Decompress stream;
    stream.input(const_cast<uint1 *>(HELLO_ZLIB), sizeof(HELLO_ZLIB));
    stream.input(const_cast<uint1 *>(WORLD_ZLIB), sizeof(WORLD_ZLIB));
    uint1 output[64] = {};
    int4 remaining = stream.inflate(output, sizeof(output));
    std::cout << "replacement.output="
              << std::string((char *)output, sizeof(output) - remaining)
              << ":remaining=" << remaining
              << ":finished=" << truth(stream.isFinished()) << '\n';
  }
  return 0;
}

File: tests/oracle/decompress_1204.rs
use rugra::compression::Decompress;
use std::ffi::CStr;

const HELLO_ZLIB: [u8; 13] = [
    0x78, 0x9c, 0xcb, 0x48, 0xcd, 0xc9, 0xc9, 0x07, 0x00, 0x06, 0x2c, 0x02, 0x15,
];
const WORLD_ZLIB: [u8; 13] = [
    0x78, 0x9c, 0x2b, 0xcf, 0x2f, 0xca, 0x49, 0x01, 0x00, 0x06, 0xa6, 0x02, 0x29,
];

fn truth(value: bool) -> &'static str {
    if value {
        "true"
    } else {
        "false"
    }
}

fn hex(bytes: &[u8]) -> String {
    bytes.iter().map(|byte| format!("{byte:02x}")).collect()
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // SAFETY: zlibVersion returns a process-lifetime NUL-terminated string.
    let version = unsafe { CStr::from_ptr(libz_sys::zlibVersion()) }.to_str()?;
    println!("zlib.version={version}");

    {
        let mut stream = Decompress::new()?;
        let mut output = [0u8; 8];
        // SAFETY: output is writable for the declared size; no input is live.
        let remaining = unsafe { stream.inflate(output.as_mut_ptr(), output.len() as i32)? };
        println!(
            "no_input.remaining={remaining}:finished={}",
            truth(stream.is_finished())
        );
    }

    {
        let mut hello = HELLO_ZLIB;
        let mut world = WORLD_ZLIB;
        let mut stream = Decompress::new()?;
        // SAFETY: both fixed arrays remain live and stable through their use.
        unsafe { stream.input(hello.as_mut_ptr(), hello.len() as i32) };
        let mut first = [0u8; 1];
        // SAFETY: hello and first remain valid and do not overlap.
        let first_remaining = unsafe { stream.inflate(first.as_mut_ptr(), first.len() as i32)? };
        // SAFETY: world remains live and stable through the next inflate call.
        unsafe { stream.input(world.as_mut_ptr(), world.len() as i32) };
        let mut second = [0u8; 64];
        // SAFETY: world and second remain valid and do not overlap.
        let second_remaining = unsafe { stream.inflate(second.as_mut_ptr(), second.len() as i32)? };
        println!(
            "midstream_replace.first_remaining={first_remaining}:second_remaining={second_remaining}:finished={}:hex={}",
            truth(stream.is_finished()),
            hex(&second[..second.len() - second_remaining as usize])
        );
    }

    {
        let mut hello = HELLO_ZLIB;
        let mut stream = Decompress::new()?;
        // SAFETY: hello remains live and stable; mutation happens between calls.
        unsafe { stream.input(hello.as_mut_ptr(), hello.len() as i32) };
        hello[0] = 0;
        debug_assert_eq!(hello[0], 0);
        let mut output = [0u8; 8];
        // SAFETY: both buffers remain valid and do not overlap during the call.
        match unsafe { stream.inflate(output.as_mut_ptr(), output.len() as i32) } {
            Ok(_) => println!("alias_mutation.error=none"),
            Err(error) => println!("alias_mutation.error={error}"),
        }
    }

    {
        let mut buffer = [0u8; 64];
        buffer[..HELLO_ZLIB.len()].copy_from_slice(&HELLO_ZLIB);
        let pointer = buffer.as_mut_ptr();
        let mut stream = Decompress::new()?;
        // SAFETY: the 64-byte allocation remains stable; zlib receives the
        // same valid 13-byte range as both input and output.
        let remaining = unsafe {
            stream.input(pointer, HELLO_ZLIB.len() as i32);
            stream.inflate(pointer, HELLO_ZLIB.len() as i32)?
        };
        println!(
            "same_address.remaining={remaining}:finished={}:hex={}",
            truth(stream.is_finished()),
            hex(&buffer[..HELLO_ZLIB.len() - remaining as usize])
        );
    }

    {
        let mut hello = HELLO_ZLIB;
        let mut stream = Decompress::new()?;
        // SAFETY: hello remains live and stable until it is fully consumed.
        unsafe { stream.input(hello.as_mut_ptr(), hello.len() as i32) };
        let mut first = [0u8; 1];
        // SAFETY: hello and first remain valid and do not overlap.
        let first_remaining = unsafe { stream.inflate(first.as_mut_ptr(), first.len() as i32)? };
        println!(
            "stream.first={}:remaining={first_remaining}:finished={}",
            String::from_utf8_lossy(&first),
            truth(stream.is_finished())
        );
        let mut second = [0u8; 64];
        // SAFETY: the unconsumed input and second remain valid and do not overlap.
        let second_remaining = unsafe { stream.inflate(second.as_mut_ptr(), second.len() as i32)? };
        println!(
            "stream.second={}:remaining={second_remaining}:finished={}",
            String::from_utf8_lossy(&second[..second.len() - second_remaining as usize]),
            truth(stream.is_finished())
        );
    }

    {
        let mut invalid = [0xde, 0xad, 0xbe, 0xef];
        let mut stream = Decompress::new()?;
        // SAFETY: invalid remains live and stable through the inflate call.
        unsafe { stream.input(invalid.as_mut_ptr(), invalid.len() as i32) };
        let mut output = [0u8; 8];
        // SAFETY: invalid and output remain valid and do not overlap.
        match unsafe { stream.inflate(output.as_mut_ptr(), output.len() as i32) } {
            Ok(_) => println!("invalid.error=none"),
            Err(error) => println!(
                "invalid.error={error}:finished={}",
                truth(stream.is_finished())
            ),
        }
    }

    {
        let mut hello = HELLO_ZLIB;
        let mut world = WORLD_ZLIB;
        let mut stream = Decompress::new()?;
        // SAFETY: each fixed array remains live and stable; the second call
        // replaces the first pointer before zlib observes it.
        unsafe {
            stream.input(hello.as_mut_ptr(), hello.len() as i32);
            stream.input(world.as_mut_ptr(), world.len() as i32);
        }
        let mut output = [0u8; 64];
        // SAFETY: world and output remain valid and do not overlap.
        let remaining = unsafe { stream.inflate(output.as_mut_ptr(), output.len() as i32)? };
        println!(
            "replacement.output={}:remaining={remaining}:finished={}",
            String::from_utf8_lossy(&output[..output.len() - remaining as usize]),
            truth(stream.is_finished())
        );
    }

    Ok(())
}
